src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal sealed class HostingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Hosting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequestsCounter;
+    private readonly Histogram<double> _requestDuration;
+
+    public HostingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-requests",
+            description: "Current requests");
+
+        _requestDuration = _meter.CreateHistogram<double>(
+            "request-duration",
+            description: "Request duration");
+    }
+
+    public void RequestStart(string scheme, string method, HostString host)
+    {
+        // Tags must match request end.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(1, tags);
+    }
+
+    public void RequestEnd(string scheme, string method, HostString host, string? routePattern, int statusCode, Exception? exception, IList<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    {
+        // Tags must match request start.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(-1, tags);
+
+        // Add information gathered during request.
+        tags.Add("status-code", GetBoxedStatusCode(statusCode));
+        if (routePattern != null)
+        {
+            tags.Add("route", routePattern);
+        }
+        // This exception is only present if there is an unhandled exception.
+        // Exceptions caught by ExceptionHandlerMiddleware and DeveloperExceptionMiddleware add exception-name to custom tags.
+        if (exception != null)
+        {
+            tags.Add("exception-name", exception.GetType().Name);
+        }
+        if (customTags != null)
+        {
+            for (var i = 0; i < customTags.Count; i++)
+            {
+                tags.Add(customTags[i]);
+            }
+        }
+
+        var duration = Stopwatch.GetElapsedTime(startTimestamp, currentTimestamp);
+        _requestDuration.Record(duration.TotalMilliseconds, tags);
+    }
+
+    public void Dispose()
+    {
+        _meter.Dispose();
+    }
+
+    public bool IsEnabled() => _currentRequestsCounter.Enabled || _requestDuration.Enabled;
+
+    private static TagList CreateRequestTags(string scheme, string method, HostString host)
+    {
+        var tags = new TagList();
+        tags.Add("scheme", scheme);
```

Comment:
Do we really want to include scheme by default as tag to this metric? While it could be useful for certain scenarios it is not useful for a majority of the cases. Speaking for services internally, all of them are `https`. This is an additional tag wasted for all these cases, and with limited number of tags for allocation free path we should be really careful.

Scheme should really be optional IMO

Reply:
It may not be useful in your case, but some apps mix http and https. Also, it's marked as a required tag by OTel.

Reply:
> This is something I would expect OTel to provide as a general purpose facility.

It has been around since the initial stable release of metrics support, here is an example https://github.com/open-telemetry/opentelemetry-dotnet/blob/b449fa1d1541341428add3b3b17cbd3c5557528b/docs/metrics/customizing-the-sdk/Program.cs#L43.

Reply:
> Its possible for libraries to also create dedicated extensibility mechanisms that control specific tags on specific instruments before they get emitted, but assuming we have mechanisms to keep allocation free either way the perf difference is likely on the order of ~10ns for an extra tag. I doubt we'd consider it worth the complexity of adding extra APIs.

+1. And not just perf, there are lots of other things that we should consider, I'll just quickly throw some on the table:
1. there are two listeners subscribing to the same meter/instrument, but they want different dimensions (e.g. one wants less dimensions to avoid too many timeseries in the TSDB, the other wants more dimensions for some fast local reaction - e.g. take a dump if xyz > 10).
2. individual libraries are most likely going to do this differently, eventually it might be hard for the user to understand and become tribal knowledge.

Here goes a similar situation https://github.com/dotnet/runtime/issues/84321#issuecomment-1497693887. My point is that it is okay to take exceptions and make one-off design - the bar should be extremely high and the ROI should be well understood.